benchmarks/persistence/persistence.py

```diff
@@ -0,0 +1,235 @@
+            a folder path as a string.
+        """
+        separator = "/" if strategy is PathStrategy.HIERARCHICAL else "_"
+
+        file_usable_date = self.execution_date_time.strftime("%Y-%m-%d")
+        components = [
+            self.benchmark_name,
+            self.branch,
+            self.latest_baybe_tag,
+            file_usable_date,
+            self.commit_hash,
+        ]
+
+        sanitized_components = [
+            self._string_sanitizer(component) for component in components
+        ]
+        path = separator.join(sanitized_components) + separator + "result.json"
+
+        return Path(path)
+
+
+class ObjectWriter(Protocol):
+    """Interface for interacting with storage."""
+
+    def write_json(self, object: dict, path_constructor: PathConstructor) -> None:
+        """Store a JSON serializable dict according to the path.
+
+        If the respective object exists, it will be overwritten.
+
+        Args:
+            object: The object to be persisted.
+            path_constructor: The PathConstructor to create the path for the object.
+        """
+
+
+@define
+class S3ObjectWriter(ObjectWriter):
+    """Class for persisting objects in an S3 bucket."""
+
+    _bucket_name: str = field(validator=instance_of(str), init=False)
+    """The name of the S3 bucket where the results are stored."""
+
+    _object_session: Session = field(factory=boto3.session.Session)
+    """The boto3 session object. This will load the respective credentials
+    from the environment variables within the container."""
+
+    @_bucket_name.default
+    def _default_bucket_name(self) -> str:
+        """Get the bucket name from the environment variables."""
+        if not PERSIST_DATA_TO_S3_BUCKET:
+            raise ValueError(
+                "No S3 bucket name provided. Please provide the "
+                + "bucket name by setting the environment variable "
+                + "BAYBE_BENCHMARKING_PERSISTENCE_PATH."
+            )
+        return os.environ["BAYBE_BENCHMARKING_PERSISTENCE_PATH"]
+
+    @override
+    def write_json(self, object: dict, path_constructor: PathConstructor) -> None:
+        """Store a JSON serializable dict according to the path.
+
+        This method will store an JSON serializable dict in an S3 bucket.
+        The S3-key of the Java Script Notation Object will be created by
+        the path object. I the key already exists, it will be overwritten.
+
+        Args:
+            object: The object to be persisted.
+            path_constructor: The PathConstructor to create the path for the object.
+
+        """
+        client = self._object_session.client("s3")
+
+        key = path_constructor.get_path(strategy=PathStrategy.HIERARCHICAL)
+
+        client.put_object(
+            Bucket=self._bucket_name,
+            Key=key.as_posix(),
+            Body=json.dumps(object),
+            ContentType="application/json",
+        )
+
+
+@define
+class LocalFileSystemObjectWriter(ObjectWriter):
```

Comment:
It came up in [this thread](https://github.com/emdgroup/baybe/pull/429#discussion_r1846198391). However, I see your point. Do you think it would be sufficient to change the logic by making the S3 Writer the default and only using local persistence when we are not using the pipeline? There are GitHub standard environment variables that could be checked.

Reply:
There is also the `CI` env that is automatically set on Github that we could leverage. That is, we could hardwire the logic that when `CI` is present, stuff automatically goes to S3. That would make it fail-safe plus we could still execute locally

Reply:
I've added that, please let me know if that is sufficient.

Reply:
To me, this seems quite solid for the current intended use 👍🏼  But I'll let @Scienfitz either resolve or add more comments